docs/internal/types.md

```diff
@@ -0,0 +1,287 @@
+ * `readonly` defines local immutability.
+   No object reached via a `readonly` capability may be modified via that the object is not guaranteed to be immutable.
+   `mut` and `readonly` capabilities to the same object may be visible at the same time.
+
+Since capabilities, unions and intersections are types themselves, an `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+A `freeze` operation consumes an `iso`, destroying the only copy of the capability that is required to mutate any object in a region, and returns an `imm` capability.
+There is no corresponding `unfreeze` operation to consume an `imm` capability and provide any kind of mutable capability.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific monotonic semantics that change the set of operations that are permitted with the resulting capability.
+
+## Module / Class model
+
+TODO
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can call a function named increment
+interface Example
+{
+  // All derived classes will have the common field
+  var common : I64;
+
+  // Classes that implement this interface must implement this method and provide the field
+  // Classes that inherit from this interface must implement this method but get the field from the base class
+  increment();
+};
+
+// This class inherits, so it gets the field `common` from the interface
+class MyType : Example
+{
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 1;
+  }
+};
+
+// This class implements the interface, so it must provide the `common` field
+class OtherType {
+  // Must provide the field
+  var common : I64;
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 2;
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class or interface and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also have a keyword `embed`, which means ...
+
+### Methods
+
+Methods are functions that belong to a class or interface which can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) or dynamically (via `object.method()`).
+On the static case, the `self` pointer is a descriptor for the type, while on the dynamic case, it's a descriptor to the instance (and its vtable, type info, etc).
```

Comment:
I think this is conflating a few things.  If I have a class like this:

```
class Cls 
{
  method() {}
}
```
The following are both valid and equivalent:

```
Cls::method();
var obj : Cls = ...;
obj.method();
```

This is just about lookup.  Any method that does not take a `Self` parameter can be invoked either by specifying the class or by specifying an object of that class.  In both cases, this is static dispatch: the compiler knows that the method that is being invoked is `Cls::method` and so will generate a direct call to that function (which will probably be inlined).

This gets more interesting with this case:

```
interface Methodic
{
  method();
};
```

Now, `Cls` implements the interface `Methodic`.  I can *also* write this:

```
var ifc : Methodic = ...;
ifc.method();
```

If the concrete type of the current value of `ifc` is an instance of `Cls` then this will call the same method as above, but now we're doing it via dynamic dispatch. Without optimisation, this will generate a vtable lookup in the vtable for the object using the selector for `method()` and will then insert the call.  With optimisation, we may discover that (especially after some inlining) we can guarantee the concrete type of `ifc` and so insert the correct call.  In the trivial case, if `Cls` is the only class that implements `method()` then we can infer this without doing any clever dataflow things.

This is almost completely orthogonal to whether there's a `Self` parameter.  If we extend `Cls` to look like this:

```
class Cls 
{
  method() {}
  instance_method(self : Self & mut) {}
}
```

Now we can invoke this as either:

```
var obj : Cls = ...;
Cls::instance_method(obj);
obj.instance_method();
```

In both cases, we need to specify the receiver (the `Self` parameter, in either procedural or an OO syntactic style).  Again, both of these will call `Cls::instance_method` directly, *without* any dynamic dispatch.

If we extend `Methodic` to also include this instance method:

```
interface Methodic
{
  method();
  instance_method(self : Self);
};
```

Now we are able to call the instance method dynamically:

```
var ifc : Methodic = ...;
ifc.instance_method();
```

This will use the selector that corresponds to `instance_method(Self)`  to look up the method in the vtable of whatever object `ifc` is.  That method will then be invoked with `ifc` as the first argument.

Reply:
Right, there are a few issues with that phrase. I didn't mean having a `self ` or not was the difference, only what descriptor was being used for. But I get that it was misleading (and also probably wrong).

IIUC, you can only modify the object is there is a self, statically or dynamically called, because otherwise, you have no base pointer on what to change.

What confuses me is where does `self` go.

```
class Foo {
  var field ...;
  create() { ... }
  method1(self : Self & mut) {
    self.field = ...;
  }
  method2() {
    // cannot touch field
    // does this mean I cannot call this as x.method2()?
  }
};
...
var x = Foo;
x.method1(); // -> method1(x);
x.method2(); // -> method2(x)??
```

Reply:
I have rewritten and  tried to give two examples. Hopefully this is clearer. 

I don't want to include the whole discussion. Although very pertinent, the document should only have what we agreed, not what we discussed.